Add pure pursuit test

Skandis opened the pull request.

mlists committed:
Robot: Automatically snap to angles when aligning, based on imu

Skandis committed:
Add pure pursuit test and add hypothesis to requirments

Skandis committed:
Add pure pursuit test and add hypothesis to requirments

Skandis committed:
Fix elif bug

Skandis committed:
Fix elif bug

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -15,3 +19,60 @@ def test_trapezoidal():
+        assert speed >= start_speed
+    if end_speed < start_speed:
+        assert speed >= end_speed - 1e-8
+        assert speed <= start_speed + 1e-8
```
Consolidate these assert statements.

auscompgeek replied:
What if `start_speed == end_speed`?

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -1,4 +1,8 @@
 import utilities.pure_pursuit as pp
+from utilities.pure_pursuit import PurePursuit
```
This is already accessible as `pp.PurePursuit`; why import this into your own namespace?

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -1,4 +1,8 @@
+from utilities.pure_pursuit import PurePursuit
+from hypothesis.strategies import floats
+from hypothesis import given
+import math
```
This seems like an odd ordering for imports... but whatever.

mlists replied:
They should go in alphabetical order divided into three sections.
Generic imports e.g. math
Related package imports e.g. hypothesis
Custom imports e.g. purepursuit

Skandis committed:
Clean up

Skandis committed:
Clean up

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -63,16 +62,35 @@ def test_distance_along_path(x):
+    PP.last_robot_y = 0
     robot_position = (pos_x, 0)
-    pure_pursuit.waypoints = [
+    PP.waypoints = [
```
Why are you assigning this directly?

Skandis replied:
What do you mean by this?

Skandis committed:
Fix assigning waypoints directly

Skandis committed:
Fix assigning waypoints directly

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -15,3 +18,77 @@ def test_trapezoidal():
+    floats(min_value=0, max_value=5),
+)
+def test_find_speed(distance_along_path, start_speed, end_speed):
+    PP = pp.PurePursuit(0.2, 0.25)
```
Is this conceptually a constant?

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -15,3 +18,77 @@ def test_trapezoidal():
+        distance_along_path,
+    )
+    if end_speed > start_speed:
+        assert end_speed + 1e-8 >= speed >= start_speed - 1e-8
```
Seems like an odd order for a range comparison.

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -15,3 +18,77 @@ def test_trapezoidal():
+def test_compute_direction(pos_x):
+    PP = pp.PurePursuit(0.2, 0.25)
+    pp.PurePursuit.last_robot_x = 0
+    pp.PurePursuit.last_robot_y = 0
```
???

auscompgeek commented on tests/test_pure_pursuit.py:
```diff
@@ -15,3 +18,77 @@ def test_trapezoidal():
+    pp.PurePursuit.last_robot_x = 0
+    pp.PurePursuit.last_robot_y = 0
+    robot_pos = (pos_x, 0)
+    waypoints = [pp.Segment(0, 0, 0, 0, 0), pp.Segment(10, 0, 0, 0, 10)]
```
Why construct this list when you're not really using it?

mlists committed:
Lower Kp of steer module

mlists committed:
Fix swerve unwinding

mlists committed:
Remove hatch intake aligner, limit switches replace it

mlists committed:
Fix aligner imports in other files

mlists committed:
Put hatch switches to networktables

mlists committed:
Add hatch alignment that uses switches

mlists committed:
Default hatch to false

james-ward committed:
Switch from driving towards lookahead point to lateral proportional control for vision alignment

james-ward committed:
Use namedtuple in vision component odometry cache

james-ward committed:
Bugfix: correct function to rotate a vector

james-ward committed:
Add test for vision component

james-ward committed:
Compensate for heading changes during alignment

james-ward committed:
Run Black on utilities/functions.py

james-ward committed:
Always return the position of the last seen fiducial, even if it is stale

james-ward committed:
Update vision processing time even when not aligning

james-ward committed:
Remove unused imports

Skandis committed:
Fix test_speed assert, fix test compute direction

Skandis committed:
Fix test_speed assert, fix test compute direction

mlists committed:
Update odometry in execute

mlists committed:
Read azimuth with a single function

store this value locally when we issue commands to steer modules

Skandis committed:
Fixed-up imports

Skandis committed:
Fixed-up imports

Skandis committed:
Run black

Skandis committed:
Run black

james-ward committed:
Proper type hinting

fizzlebert committed:
Update cargo code

- Remove heights not being used
- Use sparkmaxes for arm motor
- Use victor spx for intake motor

james-ward committed:
Remove unused import

james-ward committed:
Merge pull request #36 from james-ward/vision-latency

Vision latency improvements

mlists committed:
Replace magic numbers with an enum

mlists committed:
Remove unused variable

joelbryla committed:
Added retracting arm on climb, reduced slow down distance and reset encoder on limit switch

mlists committed:
Update to match master

mlists committed:
Move has hatch to setup

LucienMorey committed:
Merge pull request #39 from mlists/barker-changes

Barker changes

joelbryla committed:
Added climb controls

fizzlebert committed:
Merge pull request #40 from Ninjakow/climb

Climb tuning

Commit:
Change alignment automation to recognise negative fiducial range as cargo camera

Commit:
Never have less than 10% throttle

Means that robot will always move even if throttle is closed.

Commit:
When climbing, start driving forward as soon as front lift leg is above ground level

Commit:
Don't act on joystick commands when an automation is running

Commit:
Allow large joystick movements to break out of a running automation

Commit:
Add gamepad buttons for retracting climbing mech in test mode

Commit:
Fix syntax errors

james-ward committed:
Merge pull request #43 from james-ward/cleanup

Cleanup and improvement of various functions

Commit:
Cleanup logic of cargo state machine

Commit:
Avoid inadvertent outtake of cargo due to bumping gamepad triggers

Commit:
Bugfix: incorrect member variable names

Commit:
Switch cameras when running cargo mechanism

Commit:
Black cleanup

LucienMorey committed:
Merge pull request #44 from james-ward/cargo

Cargo cleanup and added functionality

Skandis committed:
Merge branch 'tests' of https://github.com/Skandis/pydeepspace into tests

auscompgeek commented:
oh no what have you done
